### ISSUE TITLE
Missing support for the new b2clogin.com domain

### ISSUE THREAD

**Issue by JohanOhlin** [AzureAD/microsoft-authentication-library-for-js#435]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[X] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Other... Please describe:
</code></pre>

## Browser:
- [X] Chrome version 68.0.3440.106
- [X] Firefox version 62.0
- [ ] IE version (not tested)
- [X] Edge version 42.17134.1.0
- [ ] Safari version (not tested)

## Library version
Library version: 0.2.3

## Current behavior
When I try to use the new recommended login domain I get the following error "endpoints_resolution_error:Endpoints cannot be resolved" and the login popup window closes right away. No outbound call is made and the error comes from within msal.js. The problem is due to that the new domain name isn't available in the TrustedHostList. Things get a little more complicated by the fact that my b2c tenant name becomes a sub domain in the new domain name, compared to before when all shared the same domain name, but I'm sure you smart guys can fix that :-)

In the sample application, this is not an issue since it's using a very old version of msal.js.

## Expected behavior
Domain should be allowed when using msal.js

**Comment by Sikta** [comment]
https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.clientapplicationbase.validateauthority?view=azure-dotnet#Microsoft_Identity_Client_ClientApplicationBase_ValidateAuthority


`Gets/sets a boolean value telling the application if the authority needs to be verified against a list of known authorities. The default value is true. It should currently be set to false for Azure AD B2C authorities as those are customer specific (a list of known B2C authorities cannot be maintained by MSAL.NET). This property can be set just after the construction of the application and before an operation acquiring a token or interacting with the STS.`

After many hours of fumbling around trying to get Azure AD B2C working from a React UI with a .NET Core 2.1 Web API, I finally got it working. The web api was requiring the B2Clogin.com url and the MSAL was only working with the login.microsoft.com url until I discovered the link above. Made the change of my authority in the MSAL config and set validateAuthority: false....finally it works!!!!!

**Comment by Mortana89** [comment]
What's the progress in this issue?

**Comment by simonneedham** [comment]
I too would like to hear what the timeline is for resolving this issue.

When a developer starts investigating & learning about Azure AD B2C they're encouraged to set everything up with the b2clogin.com domain and use  MSAL.js library for JavaScript based apps. It's kind of a crazy situation that MSAL still isn't supporting the b2clogin.com domain.

**Comment by loganfromlogan** [comment]
Looking for a resolution on this as well.

**Comment by Sikta** [comment]
Just to be clear, you can use the MSAL.js library with the b2clogin.com domain. You need to set the validateAuthority property on the client to false in order to do so.

```
const userAgentApplication = new UserAgentApplication(AzureConfig.clientId, 'https://MySubDomain.b2clogin.com/tfp/........', authCallback, {
  storeAuthStateInCookie: true,
  redirectUri: AzureConfig.redirectUri,
  postLogoutRedirectUri: AzureConfig.logoutRedirectUri,
  validateAuthority: false,
  cacheLocation: "localStorage"
});
```

**Comment by Highspeed7** [comment]
What a joke... you release a so-called authentication library and can't even keep it up to date with the latest ad b2c practices. Plus you don't respond to issues on your repository. Terrible.

**Comment by sameerag** [comment]
@Highspeed7 It is not so simple to validate B2C. It is virtually hundreds of non-fixed domains that can be aliases. I do realize that we can have a better interface and this is something which is brought up frequently in our internal discussions. We are working in a more feasible solution and will announce as and when we are ready.

Regarding response time, we are trying hard to remedy that. We are in middle of a revamp for msal-core and plan to have a better turnaround time once we have the release out. I apologize for the delay. Please reach out for any immediate blockers and we will do our best.

**Comment by Highspeed7** [comment]
@Sikta the validateAuthority: false flag is not working. I'm still getting the endpoints error.

**Comment by piotrekkmt** [comment]
+1 from me. What's the status on this?

**Comment by sameerag** [comment]
@piotrekkmt Now that MSAL core 1.0.0 release is made, we started requirement discussions with the B2C service team on this feature and will announce the dates on this support soon. I will update this ticket as and when it is finalized.

**Comment by hemantshelar** [comment]
> Just to be clear, you can use the MSAL.js library with the b2clogin.com domain. You need to set the validateAuthority property on the client to false in order to do so.
> 
> ```
> const userAgentApplication = new UserAgentApplication(AzureConfig.clientId, 'https://MySubDomain.b2clogin.com/tfp/........', authCallback, {
>   storeAuthStateInCookie: true,
>   redirectUri: AzureConfig.redirectUri,
>   postLogoutRedirectUri: AzureConfig.logoutRedirectUri,
>   validateAuthority: false,
>   cacheLocation: "localStorage"
> });
> ```

@Sikta I tried above solution but getting below error:

`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://XXX.b2clogin.com/XXX.onmicrosoft.c…thorize?p=b2c_1_signup/v2.0/.well-known/openid-configuration. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing).`

I am using '"@azure/msal-angular": "^0.1.2" library in my Angular 7 project. f Below is sample code.

```
public login(): void {
    //this.authService.loginPopup ();
    const uaa = new UserAgentApplication(
      '6b18c4cf-09f4-4eec-bcfb-d05a4a78XXXX',
      'https://XXX.b2clogin.com/XXX.onmicrosoft.com/oauth2/v2.0/authorize?p=B2C_1_Signup', (
        errDesc: string,
        token: string,
        err: string,
        tokenType: string,
        userState: string
      ) => {
        alert( token);
      },
      {
        validateAuthority: false,
        cacheLocation: 'localStorage'
      }
    );
    uaa.loginPopup(['https://XXX.onmicrosoft.com/childcare/user_impersonation']);
  }
```

**Comment by jologz** [comment]
status? I really want to use the b2clogin.  Our .Net Core app is using b2clogin but our client facing is still using login.microsoftonline.com

**Comment by sameerag** [comment]
@jologz  As mentioned above, as long as you pass 'validateAuthority: false' in the configuration for MSAL JS, b2clogin.com is supported. Can you please send us the error you are facing?

**Comment by jologz** [comment]
> @jologz As mentioned above, as long as you pass 'validateAuthority: false' in the configuration for MSAL JS, b2clogin.com is supported. Can you please send us the error you are facing?

yes it works if `validateAuthority` is set to false

**Comment by sameerag** [comment]
Great! We will change this interface in the next major release as this would be a breaking change. Until then, please set `validateAuthority` to false for all B2C use cases. We are not compromising validation in any sense since the B2C endpoint does have preconditions to be hit. It is simply impossible to white list all of the possible B2C domains.

Alternative approaches are in discussion and we will update when this changes.

**Comment by sameerag** [comment]
Closing this as we are tracking these changes with B2C changes in the next major release.

**Comment by ayeshN** [comment]
Setting `validateAuthority: false` is not working on react. Still getting the error `Access to XMLHttpRequest at 'https://login.microsoftonline.com/tfp/brainsafeb2c.onmicrosoft.com/b2c_1_brainsafe_signing_and_signup/v2.0/.well-known/openid-configuration' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource`

**Comment by lorenzobrito** [comment]
any update about this? does anyone knows if Azure has a plan to provide better documentation ?

**Comment by lorenzobrito** [comment]
> Closing this as we are tracking these changes with B2C changes in the next major release.

Still happen.

**Comment by tomgron** [comment]
The whole plethora of options and different clients and authentication options are REALLY confusing. Having this stuff well laid out for developers would be brilliant idea (e.g. Auth0 does this well). Most of the documentation is obsolete and the SPA examples in github still use some dummy jQuery stuff - come on : who uses that anymore, especially in greenfield apps. Really disappointing, going to move on to some other auth provider.

**Comment by pkanher617** [comment]
@tomgron Could you please point out where in our samples you see jQuery? We've tried to keep our documentation and samples up-to-date on our repo, which sometimes doesn't make it out to the docs.microsoft.com pages. If you have specific feedback please open an issue.

@lorenzobrito @ayeshN If you are still having problems with this please review the currently open issues around the B2C CORS errors and open a new issue if you don't find one that matches yours.
